Improve dependencies task output documentation (#22456)
gradle/gradle: subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/viewing_debugging_dependencies.adoc

[[viewing-debugging-dependencies]]
[[sec:debugging-build-scans]]
= View and Debug Dependencies

Gradle provides tooling to navigate dependency graphs and mitigate link:https://en.wikipedia.org/wiki/Dependency_hell[dependency hell].
Users can render the full graph of dependencies as well as identify the selection reason and origin for a dependency.
Dependencies can originate through build script declared dependencies or transitive dependencies.

> You can visualize dependencies with the built-in Gradle CLI `dependencies` task or with link:https://scans.gradle.com/[build scans].

Reviewer: > ... with the built-in Gradle CLI `dependencies` and `dependencyInsight` tasks ...

Let's mention all the things this page covers in the intro.

A project may request two different versions of the same dependency either directly or transitively.
Gradle applies <<dependency_resolution.adoc#sec:version-conflict,version conflict resolution>> to ensure that only one version of the dependency exists in the dependency graph.
The following example introduces a conflict with `commons-codec:commons-codec`, added both as a direct dependency and a transitive dependency of JGit:

====
include::sample[dir="snippets/dependencyManagement/inspectingDependencies-dependencyInsightReport/groovy",files="build.gradle[tags=dependency-declaration]"]
include::sample[dir="snippets/dependencyManagement/inspectingDependencies-dependencyInsightReport/kotlin",files="build.gradle.kts[tags=dependency-declaration]"]
====

The dependency tree in a link:https://scans.gradle.com/[build scan] shows information about conflicts.
Click on a dependency and select the "Required By" tab to see the selection reason and origin of the dependency.

image::dependency-management-dependency-insight-report-build-scan.png[]

=== Dependency Insights

Gradle provides the built-in `dependencyInsight` task to render a  _dependency insight report_ from the command line.

> Dependency insights only provide information about a single project at a time.

Reviewer: > ... only provide information about a single configuration and single project ...

Let's state both elements, since we covered them for `dependencies`

Given a dependency, you can identify the selection reason and origin.

`dependencyInsight` accepts the following parameters:

`--dependency <dependency>` (mandatory)::
The dependency to generate an insight report for.
You can supply a complete `group:name`, or part of it.
If multiple dependencies match, Gradle generates a report covering all matching dependencies.
`--configuration <name>` (mandatory)::
The dependency configuration from which to supply the dependency information.
This parameter is optional for projects that use the <<java_plugin#java_plugin, Java plugin>>, since the plugin provides a default value of `compileClasspath`.
`--single-path` (optional)::
Render only a single path to the dependency.

> ----

Reviewer: This example, without an Example header, feels out of place.

Why not move it after the "Selection reasons" section, since it illustrates both elements: the command and the output.

Gradle offers both capabilities through visual representation via build scans and as command line tooling.

[[sec:debugging-build-scans]]
== Build scans
= View and Debug Dependencies

[NOTE]
====
If you do not know what link:https://scans.gradle.com/[build scan]s are, be sure to check them out!
====
Gradle provides tooling to navigate dependency graphs and mitigate link:https://en.wikipedia.org/wiki/Dependency_hell[dependency hell].
Users can render the full graph of dependencies as well as identify the selection reason and origin for a dependency.
Dependencies can originate through build script declared dependencies or transitive dependencies.
You can visualize dependencies with the built-in Gradle CLI `dependencies` task or with link:https://scans.gradle.com/[build scans].

A build scan can visualize dependencies as a navigable, searchable tree.
Additional context information can be rendered by clicking on a specific dependency in the graph.
[[sec:listing_dependencies]]
== List Project Dependencies

.Dependency tree in a build scan

> image::dependency-management-dependencies-report-build-scan.png[]

Reviewer: I would keep the build scan example but move it close to the output example for `dependencies`.

Author: I feel that leaving the image distracts from the core idea behind the page: explaining the `dependencies` task. Since we link to other documentation that explains the ins and outs of build scans, I think it's OK to lose this example.

I would be comfortable adding the example back in if we elaborated a bit more on how to use the build scan tree for real world debugging use cases. Maybe there's another section of the documentation we could consolidate here? I'll look, but if you know of any, you could save me some time.

Reviewer: My comment came from the fact that build scan example for `dependencyInsight` has been preserved.

Like I said, the flow of the page is much better with your changes. So if you think having it lower down the page is too much noise, good for me.

> - `(c)`: Dependencies whose version is limited by dependency constraints.

Reviewer: The annotation comment is wrong.

Here is the full explanation:

Indicates that the element is a dependency constraint. The representation never follows transitives from a constraint. Look for the matching dependency elsewhere in the tree.

Author: Brilliant, this makes much more sense to me. I'll also add this to the CLI output because the current description just says "dependency constraint" with no additional info.

Author: Is this a more accurate reword?
```suggestion
- `(c)`: A dependency constraint directly limits the version of this element. Does not apply to indirect constraints introduced transitively through another element's constraint.
```

Author: Fixed!

> - `(n)`: A dependency or dependency configuration that cannot be resolved. Sometimes called a producer.

Reviewer: > Sometimes called a producer.

This reads wrong. What do you want to say?

Non resolvable configurations are either dependency buckets or consumable ones. 
And for these last two, the report will still print dependencies declared directly on the configuration

Author: Is this more accurate?

```suggestion
- `(n)`: A dependency or dependency configuration that cannot be resolved. Sometimes called _unresolvable_.
```

Reviewer: Not really, as we do not use that term anywhere AFAIK. What could be done is link to the section explaining the different kinds of configurations.

Author: Fixed!

====
include::sample[dir="snippets/dependencyManagement/inspectingDependencies-dependencyInsightReport/groovy",files="build.gradle[tags=dependency-declaration]"]
include::sample[dir="snippets/dependencyManagement/inspectingDependencies-dependencyInsightReport/kotlin",files="build.gradle.kts[tags=dependency-declaration]"]
====

The dependency tree in a link:https://scans.gradle.com/[build scan] shows information about conflicts.
Click on a dependency and select the "Required By" tab to see the selection reason and origin of the dependency.

image::dependency-management-dependency-insight-report-build-scan.png[]

=== Dependency Insights

Gradle provides the built-in `dependencyInsight` task to render a  _dependency insight report_ from the command line.
Dependency insights only provide information about a single project at a time.
Given a dependency, you can identify the selection reason and origin.

`dependencyInsight` accepts the following parameters:

`--dependency <dependency>` (mandatory)::
The dependency to generate an insight report for.
You can supply a complete `group:name`, or part of it.

> If multiple dependencies match, Gradle generates a report for each matching dependency.

Reviewer: It is not a report per dependency, it is a single report which covers multiple dependencies.

Author: Is this more accurate?

```suggestion
If multiple dependencies match, Gradle generates a report covering all matching dependency.
```

Reviewer: Yes

You can visualize dependencies with:

A build scan can visualize dependencies as a navigable, searchable tree.
Additional context information can be rendered by clicking on a specific dependency in the graph.

.Dependency tree in a build scan
image::dependency-management-dependencies-report-build-scan.png[]
- the built-in Gradle CLI `dependencies` task
- the built-in Gradle CLI `dependencyInsight` task
- link:https://scans.gradle.com/[build scans]

[[sec:listing_dependencies]]
== Listing dependencies in a project
== List Project Dependencies

Gradle can visualize the whole dependency tree for every <<declaring_dependencies.adoc#sec:what-are-dependency-configurations,configuration>> available in the project.
Gradle provides the built-in `dependencies` task to render a dependency tree from the command line.
By default, the dependency tree renders dependencies for all <<declaring_dependencies.adoc#sec:what-are-dependency-configurations,configurations>>, but only a <<command_line_interface#executing_tasks_in_multi_project_builds,single project>>.
The dependency tree indicates the selected version of each dependency.
It also displays information about dependency conflict resolution.

> Rendering the dependency tree is particularly useful if you'd like to identify which dependencies have been resolved at runtime.

Reviewer 2: I think this section should emphasize the distinction between the declared dependencies in a build file and the transitive deps that are resolved, which the report makes visible and are almost always the source of problems.

> gradle -q dependencies --configuration tRC
----

> To see a list of all the pre-defined configurations added by the `java` plugin, see <<java_plugin.adoc#sec:java_plugin_and_dependency_management,the documentation for the Java Plugin>>.

Reviewer 2: My thinking here is to not seem to exclude Groovy or Kotlin projects.

> To see a list of all the pre-defined configurations added by the `java` plugin, see <<java_plugin.adoc#sec:java_plugin_and_dependency_management,the documentation for the Java Plugin>>.

Author: ```suggestion
To see a list of all the configurations available in a project, including those added by any plugins, you can run a `resolvableConfigurations` report.

For more info, see that plugin's documentation (for instance, the Java Plugin is documented <<java_plugin.adoc#sec:java_plugin_and_dependency_management,here>>).
```

A project may request two different versions of the same dependency either directly or transitively.
Gradle applies <<dependency_resolution.adoc#sec:version-conflict,version conflict resolution>> to ensure that only one version of the dependency exists in the dependency graph.
The following example introduces a conflict with `commons-codec:commons-codec`, added both as a direct dependency and a transitive dependency of JGit:

====
include::sample[dir="snippets/dependencyManagement/inspectingDependencies-dependencyInsightReport/groovy",files="build.gradle[tags=dependency-declaration]"]
include::sample[dir="snippets/dependencyManagement/inspectingDependencies-dependencyInsightReport/kotlin",files="build.gradle.kts[tags=dependency-declaration]"]
====

The dependency tree in a link:https://scans.gradle.com/[build scan] shows information about conflicts.
Click on a dependency and select the "Required By" tab to see the selection reason and origin of the dependency.

image::dependency-management-dependency-insight-report-build-scan.png[]

=== Dependency Insights

Gradle provides the built-in `dependencyInsight` task to render a  _dependency insight report_ from the command line.

> Dependency insights provide information about a single <<declaring_dependencies.adoc#sec:what-are-dependency-configurations,configuration>> within a single <<command_line_interface#executing_tasks_in_multi_project_builds,project>>.

Author: ```suggestion
Dependency insights provide information about a single dependency within a single <<declaring_dependencies.adoc#sec:what-are-dependency-configurations,configuration>>.
```

Given a dependency, you can identify the selection reason and origin.

`dependencyInsight` accepts the following parameters:

`--dependency <dependency>` (mandatory)::
The dependency to generate an insight report for.
You can supply a complete `group:name`, or part of it.
If multiple dependencies match, Gradle generates a report covering all matching dependencies.
`--configuration <name>` (mandatory)::
The dependency configuration from which to supply the dependency information.
This parameter is optional for projects that use the <<java_plugin#java_plugin, Java plugin>>, since the plugin provides a default value of `compileClasspath`.
`--single-path` (optional)::
Render only a single path to the dependency.

> The following code snippet demonstrates how to run a dependency insight report for a all paths to a dependency named "commons-codec" within the "scm" configuration:

Reviewer 2: ```suggestion
The following code snippet demonstrates how to run a dependency insight report for all paths to a dependency named "commons-codec" within the "scm" configuration:
```

----
> gradle -q dependencyInsight --dependency commons-codec --configuration scm
include::{snippetsPath}/dependencyManagement/inspectingDependencies-dependencyInsightReport/tests/dependencyInsightReport.out[]
----

For more information about configurations, see the <<declaring_dependencies.adoc#sec:what-are-dependency-configurations,dependency configuration documentation>>.

==== Selection Reasons

The "Selection reasons" section of the dependency insight report lists the reasons why a dependency was selected.
Have a look at the table below to understand the meaning of the different terms used:

.Terminology
[%header%autowidth,compact]
|===
| Reason    | Meaning

| (Absent)
| No reason other than a reference, direct or transitive, was present.

| Was requested : <text>

> | The dependency appears and the inclusion came with a <<declaring_dependencies#sec:documenting-dependencies, `because` text>>.

Author: ```suggestion
| The dependency appears in the graph, and the inclusion came with a <<declaring_dependencies#sec:documenting-dependencies, `because` text>>.
```

[[sec:resolving-variant-aware-errors]]
=== Variant Selection Errors

Sometimes a selection error happens at the <<variant_model.adoc#understanding-variant-selection,variant selection level>>.
Have a look at the <<variant_model.adoc#sec:variant-select-errors,dedicated section>> to understand these errors and how to resolve them.

[[sub:resolving-unsafe-configuration-resolution-errors]]
=== Unsafe Configuration Resolution Errors

Resolving a configuration can have side effects on Gradle's project model.
As a result, Gradle must manage access to each project's configurations.
There are a number of ways a configuration might be resolved unsafely.
For example:

* A task from one project directly resolves a configuration in another project in the task's action.
* A task specifies a configuration from another project as an input file collection.
* A build script for one project resolves a configuration in another project during evaluation.
* Project configurations are resolved in the settings file.

Gradle produces a deprecation warning for each unsafe access.
Unsafe access can cause indeterminate errors.

> You should fix unsafe access warnings in your build.

Author: ```suggestion
You should <<command_line_interface.adoc#sec:command_line_warnings,fix unsafe access warnings>> in your build.
```
